### ISSUE TITLE
PasswordGrant relying on UserRepositoryInterface

### ISSUE THREAD

**Issue by PhilippSchaffrath** [thephpleague/oauth2-server#679]
**TLDR**: `UserRepositoryInterface` should be called `UserProviderInterface`.

The `PasswordGrant` uses an object implementing the `UserRepositoryInterface` to validate the user.
In cases where the user's password is stored in an encoded (hashed) form, you would have to handle that logic in the custom `UserRepository`. But this is logic that shouldn't live inside a `Repository`.

Writing a `UserProvider` that implements `UserRepositoryInterface` is possible and would solve that problem, but kind of breaks the domain here.

Instead of providing the `UserRepositoryInterface`, i would much rather see a `UserProviderInterface`. That way it's completely up to the implementation on how to handle things.

I know, that's really first world problems, but it just tickles that one part in my brain...

**Comment by alexbilbie** [comment]
I understand what you're saying about the naming convention however I don't understand this comment - "That way it's completely up to the implementation on how to handle things".

What _can't_ you currently do that you could do if the interface was renamed? Or is this purely a semantic change?

**Comment by PhilippSchaffrath** [comment]
It's purely semantic, not really a bug or anything. Currently running a `UserProvider` that implements the `UserRepositoryInterface` which works, but isn't nice.
